src/Sentry/Internal/BackpressureMonitor.cs

```diff
@@ -0,0 +1,166 @@
+    internal long LastRateLimitEventTicks => Interlocked.Read(ref _lastRateLimitEvent);
+
+    public BackpressureMonitor(IDiagnosticLogger? logger, ISystemClock? clock = null, bool startImmediately = true)
+    {
+        _logger = logger;
+        _clock = clock ?? SystemClock.Clock;
+
+        if (startImmediately)
+        {
+            _logger?.LogDebug("Starting BackpressureMonitor.");
+            WorkerTask = Task.Run(() => DoWorkAsync(_cts.Token));
+        }
+        else
+        {
+            WorkerTask = Task.CompletedTask;
+        }
+    }
+
+    /// <summary>
+    /// For testing purposes only. Sets the downsample level directly.
+    /// </summary>
+    internal void SetDownsampleLevel(int level)
+    {
+        Interlocked.Exchange(ref _downsampleLevel, level);
+    }
+
+    internal void IncrementDownsampleLevel()
+    {
+        var oldValue = _downsampleLevel;
+        if (oldValue < MaxDownsamples)
+        {
+            var newValue = oldValue + 1;
+            if (Interlocked.CompareExchange(ref _downsampleLevel, newValue, oldValue) == oldValue)
+            {
+                _logger?.LogDebug("System is under pressure, increasing downsample level to {0}.", newValue);
+            }
+        }
+    }
+
+    /// <summary>
+    /// A multiplier that can be applied to  the SampleRate or TracesSampleRate to reduce the amount of data sent to
+    /// Sentry when the system is under pressure.
+    /// </summary>
+    public double DownsampleFactor
+    {
+        get
+        {
+            var level = _downsampleLevel;
+            return 1d / (1 << level);
+        }
+    }
+
+    public void RecordRateLimitHit(DateTimeOffset when) => Interlocked.Exchange(ref _lastRateLimitEvent, when.Ticks);
+
+    public void RecordQueueOverflow() => Interlocked.Exchange(ref _lastQueueOverflow, _clock.GetUtcNow().Ticks);
+
+    private async Task DoWorkAsync(CancellationToken cancellationToken)
+    {
+        try
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                DoHealthCheck();
+
+                await Task.Delay(TimeSpan.FromSeconds(CheckIntervalInSeconds), cancellationToken).ConfigureAwait(false);
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Task was cancelled, exit gracefully
+        }
+    }
+
+    internal void DoHealthCheck()
+    {
+        if (IsHealthy)
+        {
+            var previous = Interlocked.Exchange(ref _downsampleLevel, 0);
+            if (previous > 0)
+            {
+                _logger?.LogDebug("System is healthy, resetting downsample level.");
+            }
+        }
+        else
+        {
+            IncrementDownsampleLevel();
+        }
+    }
+
+    /// <summary>
+    /// Checks for any recent queue overflows or any rate limit events since the last check.
+    /// </summary>
+    /// <returns></returns>
+    internal bool IsHealthy
```

Comment:
thought: method rather than property

I'm wondering if this should be a method instead of a property.
As property, I would expect it to not be time-dependent.

But since it's internal, it may also be more of a stylistic choice.
I'm also fine with keeping it a Property.

Reply:
I've left this as a property (given the name, I think it would be expected that it may return true or false at different times).